### DISCUSSION TITLE
Experiencing flaky behaviour when doing Composer lookups from custom repos

### DISCUSSION THREAD

**Discussion by herndlm** [renovatebot/renovate#23063] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.147.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

We are using multiple custom repos in our composer.json like
```json
    "repositories": [
        {
            "type": "composer",
            "url": "<repo url a>"
        },
        {
            "type": "composer",
            "url": "<repo url b>"
        },
```
and composer would look up packages in a, then b and finally fall back to packagist. This was also working fine with renovate. But recently (not sure, maybe since 1-3 weeks) I noticed that packages from repo b cannot be looked up all the time. It's very flaky and I couldn't figure out yet why it would be like that.

I just ran the same renovate version on the same server mutliple times. Once it works like this
```
 INFO: Dependency extraction complete (repository=<repo>, baseBranch=develop)
         "stats": {
           "managers": {
             "composer": {"fileCount": 1, "depCount": 15},
             "npm": {"fileCount": 1, "depCount": 29}
           },
           "total": {"fileCount": 2, "depCount": 44}
         }
  DEBUG: Dependency roave/security-advisories has unsupported/unversioned value dev-master (versioning=composer) (repository=<repo>)
  DEBUG: PackageFiles.add() - Package file saved for base branch (repository=<repo>, baseBranch=develop)
  DEBUG: Package releases lookups complete (repository=<repo>, baseBranch=develop)
  DEBUG: branchifyUpgrades (repository=<repo>)
  DEBUG: 12 flattened updates found: <package-from-repo-b>, phpunit/phpunit, react, react-dom, react-tooltip, styled-components, @testing-library/react, @types/jest, @wordpress/scripts, jest (repository=<repo>)
  DEBUG: Returning 9 branch(es) (repository=<repo>)
```

And another time it failed like this
```
INFO: Dependency extraction complete (repository=<repo>, baseBranch=develop)
         "stats": {
           "managers": {
             "composer": {"fileCount": 1, "depCount": 15},
             "npm": {"fileCount": 1, "depCount": 29}
           },
           "total": {"fileCount": 2, "depCount": 44}
         }
  DEBUG: GET https://packagist.org/p2/<package-from-repo-b>.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=304) (repository=<repo>)
  DEBUG: Datasource 404 (repository=<repo>)
         "datasource": "packagist",
         "packageName": "<package-from-repo-b>",
         "url": "https://packagist.org/p2/<package-from-repo-b>.json
  DEBUG: Failed to look up packagist package <package-from-repo-b> (repository=<repo>, packageFile=composer.json, dependency=<package-from-repo-b>)
  DEBUG: GET https://packagist.org/p2/<package-from-repo-b>~dev.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=371) (repository=<repo>)
  DEBUG: Dependency roave/security-advisories has unsupported/unversioned value dev-master (versioning=composer) (repository=<repo>)
  DEBUG: PackageFiles.add() - Package file saved for base branch (repository=<repo>, baseBranch=develop)
  DEBUG: Package releases lookups complete (repository=<repo>, baseBranch=develop)
  DEBUG: branchifyUpgrades (repository=<repo>)
  DEBUG: 11 flattened updates found: phpunit/phpunit, react, react-dom, react-tooltip, styled-components, @testing-library/react, @types/jest, @wordpress/scripts, jest (repository=<repo>)
```

According to logs it looks like it wrongfully tries to find the package on packagist.org, but I'm not sure if really that is the problem or might there be some other underlying thing that is not logged? I also have no idea why it would be so flaky..

While comparing both "http statistics" sections of the logs I can also see that repo a (the non-flaky one) and packagist.org are part of it in both. repo b (the flaky one) exists only in the working case.

Any ideas, or did something in that regard recently change maybe?

Did no create a minimal producer yet since I have the feeling this might be something on our side, but I don't fully get it yet and can look into creating one if you don't have any other ideas.

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by rarkins** [comment]
Do you see any information in `registryUrls` in the logs?

**Reply by herndlm** [reply]
yeah, and it looks fine I guess, like this:
```json
{
    "depType": "require",
    "depName": "<package-from-repo-b>",
    "currentValue": "^6.0",
    "lockedVersion": "6.1.6",
    "datasource": "packagist",
    "registryUrls": [
        "<repo a>",
        "<repo b>",
        "https://packagist.org
    ],
    "updates": [],
    "packageName": "<package-from-repo-b>",
    "versioning": "composer",
    "warnings": [
        {
        "topic": "<package-from-repo-b>",
        "message": "Failed to look up packagist package <package-from-repo-b>"
        }
    ]
},
```
`"https://packagist.org` is really missing the trailing `"` in the logs, no clue why.

**Reply by herndlm** [reply]
in comparison an entry from repo a from the same run:
```json
{ 
    "depType": "require", 
    "depName": "<package-from-repo-a>", 
    "currentValue": "^0.4.14", 
    "lockedVersion": "0.4.15", 
    "datasource": "packagist", 
    "registryUrls": [ 
        "<repo a>", 
        "<repo b>", 
        "https://packagist.org 
    ], 
    "updates": [], 
    "packageName": "<package-from-repo-a>", 
    "versioning": "composer", 
    "warnings": [], 
    "registryUrl": "<repo a>", 
    "currentVersion": "0.4.15", 
    "fixedVersion": "0.4.15" 
},
```

Repo a and repo b are using completely different composer repo software and IIRC one is using a quite old format. not sure if that might be related.

**Reply by rarkins** [reply]
So the order of registryUrls is correct from your perspective?

Another field to watch is the `registryUrl` one. That tells you which one it was looked up from

**Reply by herndlm** [reply]
I think the order should be fine. It's the same as in the composer.json (repo a, repo b, packagist). `registryUrl` is missing for package b completely (is that the issue? Any hints where I could start debugging this in the code maybe?) and what is so misleading to me is the error that it couldn't be looked up on packagist, as if it doesn't even try to check repo a or b. But for package a it worked fine.

**Reply by rarkins** [reply]
Can you check the package cache statistics from the end of each job? It will tell you how many requests to the cache succeeded or were updated

**Reply by herndlm** [reply, marked as answer]
it seems to be working stable again 🤷‍♂️
Maybe repo b or the network around it is somehow in a bad state. Because of a weird corporate network setup a proxy is also involved.. I'll try to get more logs when this happens again.
